google/cloud/pubsub/blocking_publisher_connection.cc

```diff
@@ -41,11 +62,27 @@ std::shared_ptr<BlockingPublisherConnection> MakeBlockingPublisherConnection(
   auto background = internal::MakeBackgroundThreadsFactory(opts)();
   auto stub =
       pubsub_internal::MakeRoundRobinPublisherStub(background->cq(), opts);
-  return std::make_shared<pubsub_internal::BlockingPublisherConnectionImpl>(
-      std::move(background), std::move(stub), std::move(opts));
+  return BlockingConnectionFromDecoratedStub(
+      std::move(opts), std::move(background), std::move(stub));
 }
 
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
 }  // namespace pubsub
+
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+std::shared_ptr<pubsub::BlockingPublisherConnection>
+MakeTestBlockingPublisherConnection(
+    Options const& opts, std::vector<std::shared_ptr<PublisherStub>> mocks) {
+  auto background = internal::MakeBackgroundThreadsFactory(opts)();
+  auto stub = MakeTestPublisherStub(background->cq(), opts, std::move(mocks));
+  return pubsub::BlockingConnectionFromDecoratedStub(
+      std::move(opts), std::move(background), std::move(stub));
+}
```

Comment:
Should this be in some testonly file?

Reply:
I think we'd have to move `BlockingConnectionFromDecoratedStub` to the pubsub namespace to move this to a test file. 

I moved it into the pubsub_testing namespace since I think that's a better fit.

Do you feel strongly about it being in a testonly file?

Reply:
Strongly?  I guess not.  But certainly a preference (not that I grok the problem you foresee).

I do note that `git grep -l 'namespace pubsub_testing'` currently only catches files in google/cloud/pubsub/testing.

Reply:
It is showing pubsub/* for me, might be a delay in the PR.

```
> git grep -l 'namespace pubsub_testing'
google/cloud/pubsub/blocking_publisher_connection.cc
google/cloud/pubsub/blocking_publisher_connection.h
google/cloud/pubsub/testing/fake_streaming_pull.cc
google/cloud/pubsub/testing/fake_streaming_pull.h
google/cloud/pubsub/testing/mock_batch_sink.h
google/cloud/pubsub/testing/mock_message_batch.h
google/cloud/pubsub/testing/mock_publisher_stub.h
google/cloud/pubsub/testing/mock_subscriber_stub.h
google/cloud/pubsub/testing/mock_subscription_batch_source.h
google/cloud/pubsub/testing/mock_subscription_message_source.h
```

I think this makes sense for now, it is also what the non-blocking publisher does.

I'm not sure how to do this without exposing `BlockingConnectionFromDecoratedStub` which I don't think should be a public function.

Reply:
> It is showing pubsub/* for me, might be a delay in the PR.

Yeah ... I believe those two `blocking_publisher_connection.*` matches are new from this PR, hence the question.

> I think this makes sense for now, it is also what the non-blocking publisher does.

I see that now, although the non-blocking one is in the `pubsub_internal` namespace.

> I'm not sure how to do this without exposing `BlockingConnectionFromDecoratedStub` which I don't think should be a public function.

`BlockingConnectionFromDecoratedStub()` could be a `pubsub_internal` function, usable by both the normal and testing libraries.

All that said, I don't want to delay you, so I'll approve, but it is usually best to get such things just right from the get-go.

Reply:
The refactoring seems to be quite involved. I think I'd want to also refactor the non blocking one. 

Since they are both in non-user facing interfaces it should be ok to do a refactoring...

Tracking here https://github.com/googleapis/google-cloud-cpp/issues/12970